### PR TITLE
`torch.prod` backward for complex types.

### LINKED ISSUE #53511: Add complex autograd support for torch.prod
```
>>> a = torch.randn(1, 3, dtype=torch.cdouble, requires_grad=True)
>>> torch.prod(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: prod does not support automatic differentiation for outputs with complex dtype.
```

cc @ezyang @anjali411 @dylanbespalko @mruberry @aocsa

### PR DESCRIPTION
**Pull request by nikitaved**
Fixes https://github.com/pytorch/pytorch/issues/53511
torch.det does depend on torch.prod, which in turn depends on several other functions, and they also depend on torch.prod, so there is a circular relationship, hence this PR will enable complex backward support for several functions at once.

### CODE REVIEW

**Comment by nikitaved** [comment]
@mruberry , could you please help me with this PR since @anjali411 is on PTO?

The backward for `torch.prod` does depend on `torch.index_select`, for example. When I enabled the complex backward support, I noticed that `complex_test` from `test/test_autograd.py` only works with mappings of the form `Tensor -> Tensor`, which means that the tests listed in `method_tests` will most likely fail. And, indeed, I get this:
```
        elif isinstance(arg, torch.Tensor):
            if arg.dtype == torch.float:
                arg = arg.double()
            if arg.dtype == torch.cfloat:
                arg = arg.to(torch.cdouble)
            if arg.is_complex() != dtype.is_complex:
                raise RuntimeError("User provided tensor is real for a test that runs with complex dtype, ",
>                                  "which is not supported for now")
E               RuntimeError: ('User provided tensor is real for a test that runs with complex dtype, ', 'which is not supported for now')

torch/testing/_internal/common_methods_invocations.py:1564: RuntimeError

```
I will probably ignore backward tests for `index_select` for now, but shall I write separate tests for complex inputs somewhere?

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:396]: What is the precision implication of flipping this order? I have https://github.com/pytorch/pytorch/issues/43414 in mind

File: torch/testing/_internal/common_methods_invocations.py
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:959]: You actually disable the non-complex tests as well here? That does not sound good

File: torch/testing/_internal/common_methods_invocations.py
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:1019]: Should this be a new test to handle complex to keep both the double and cdouble tests?

File: torch/testing/_internal/common_methods_invocations.py
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:959]: It is still WIP, I need `index_add` complex dispatch on CUDA for this to work.

File: torch/testing/_internal/common_methods_invocations.py
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:1019]: `double` test works only for non-complex types. Yeah, I see there a list that keeps only complex tests. I will add it there.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:396]: I am not sure they are comparable. The reason being one order is better then `input[i]` is greater than 1, the other when `input[i]` is less than 1. And it is only entry-wise dependent...

Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:396]: The different in the other one was that we were squaring the denominator before doing the division. BUt since there is no squaring it should be fine here right?

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:396]: But if you feel more comfortable with the original order, I could make it like
`grad * result.conj() / input.conj()`

Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:396]: I think the new one will work. It just that it reminded me of that issue and I didn't want to introduce any regression.

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:396]: I think you are right. I will restore the previous order so that the change does not become bc-breaking.

**Comment by anjali411** [comment]
@nikitaved sorry I was on PTO when this PR was created and didn't see it until now. is this ready for review? also can you rebase the PR on latest master?

**Comment by nikitaved** [comment]
@anjali411 , thank you for your message! It is still WIP and it is blocked by https://github.com/pytorch/pytorch/issues/48230 for now ;(

**Comment by anjali411** [comment]
@nikitaved this should be unblocked after `index_fill_` is ported right?

**Comment by nikitaved** [comment]
Appears so, I cannot reproduce test failures with ported index_fill.

**Comment by anjali411** [comment]
> Appears so, I cannot reproduce test failures with ported index_fill.

great! `index_fill_` PR is merged now. let me know when this PR is ready for review

**Comment by anjali411** [comment]
Hey @nikitaved are you planning to work on this PR?

**Comment by nikitaved** [comment]
Hey, @anjali411 , yes. Let me make my `eig_backward` PR ready first.

File: torch/testing/_internal/common_methods_invocations.py
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:3165]: @anjali411 , @mruberry , this cast test is causing issues as it does not work on master anymore. I am not sure it is reproducible with OpInfo, or am I wrong?

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py]: @nikitaved I think you need to update the formula for `cumprod` to make it work for complex tensors.

File: tools/autograd/gen_variable_type.py
Comment (anjali411) [inline tools/autograd/gen_variable_type.py:95]: there's an open PR for cumsum so I think we can exclude the cumsum related changes for this PR https://github.com/pytorch/pytorch/pull/53240

File: torch/testing/_internal/common_methods_invocations.py
Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py]: @Lezcano is working to add complex support for `torch.cumprod` so I think we can exclude `cumprod` related changes from this PR.

File: tools/autograd/gen_variable_type.py
Comment (nikitaved) [reply tools/autograd/gen_variable_type.py:95]: There is no complex prod backward without cumsum complex support. Let' see who's PR gets landed first and then one of us could merge?

Comment (nikitaved) [reply tools/autograd/gen_variable_type.py:95]: OH, I see that `cumsum` does not need any modification when it comes to autograd. Agree, we could skip it here.

File: torch/testing/_internal/common_methods_invocations.py
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py]: There is no complex prod backward without `cumprod` complex support. Let' see who's PR gets landed first and then one of us could merge?

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py]: @anjali411 , the formula for complex `cumprod_backward` is correct, the casting test is failing because it complains that `cumprod` receives and input with `self.dtype != dtype`.

File: torch/testing/_internal/common_methods_invocations.py
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3119]: let's remove these method tests and add an `OpInfo` instead

File: torch/testing/_internal/common_methods_invocations.py
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:3155]: please undo this change

File: torch/testing/_internal/common_methods_invocations.py
Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py]: this issue is similar to https://github.com/pytorch/pytorch/pull/53240#discussion_r588326560 let's just leave that method test for now and port the rest

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py]: I will write the `OpInfo`  tests for `cumprod`  and will add the complex support as well in my PR. I will write the benchmarks and push it later today if I don't encounter any major error when porting the tests to `OpInfo`.

As @nikita said, I'd also be happy to merge nikitaved's PR if his is merged first, as the overlapping is minimal.

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py]: I could write tests for `prod` only, so this way we introduce no conflicts whatsoever.

Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py]: For what is worth, complex support for `cumprod` unsurprisingly depends on `prod` (among other operations). I guess this PR will have to go in first, and then I'll we can do complex `cumprod`.

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py]: Yes, but we can decide who writes the tests for `cumprod`. I could enable its complex support, yet leave the tests intact.

File: torch/testing/_internal/common_methods_invocations.py
Comment (nikitaved) [inline torch/testing/_internal/common_methods_invocations.py:1808]: @mruberry , could you please tell whether it is something expect? I hit very weird error messages if these tests are not skipped, that is:
`consistency_jit_float16` causes `allclose` to fail, not matter what  precision I set.
`test_out` for `float32` fails with an error message "User warning not triggered".

**Comment by nikitaved** [comment]
@anjali411 , unless the skips in OpInfos are real issues, the PR is ready.

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1888]: The first test is OK to skip and I have a PR coming that will fix and remove the skip, so no worries.

The second error is probably real... looking...

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1888]: The error appears to be real. torch.prod() does not exhibit the correct behavior when given an out tensor with an incorrect size. It's OK to keep the skips but we should file an issue and comment a link to the issue here. We should fix its out= behavior in the future.

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:1888]: Strange that it only manifests itself with `float32`, which is weird.

Comment (mruberry) [reply torch/testing/_internal/common_methods_invocations.py:1888]: It's just because the test only runs in float32.

File: aten/src/ATen/native/ReduceOps.cpp
Comment (anjali411) [inline aten/src/ATen/native/ReduceOps.cpp:296]: I think you also need to modify this: https://github.com/pytorch/pytorch/blob/33aaea912abbe22fb6518807b5792d1c6bdc6829/torch/csrc/autograd/FunctionsManual.cpp#L422 to `return grad * (exclusive_normal * exclusive_reverse).conj();`?

**Comment by mruberry** [comment]
Reopening. This was not merged (the system incorrectly thinks it was).

**Comment by aocsa** [comment]
it was solved here #53854

File: aten/src/ATen/native/ReduceOps.cpp
Comment (nikitaved) [reply aten/src/ATen/native/ReduceOps.cpp:296]: These changes are already in the PR.

File: torch/csrc/autograd/FunctionsManual.cpp
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:422]: @anjali411 , here are changes, so they are already included.

File: torch/testing/_internal/common_methods_invocations.py
Comment (anjali411) [inline torch/testing/_internal/common_methods_invocations.py:1357]: shouldn't we pass the device to `prod_single_zero` here?

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:1357]: oh I see you are calling into the function defined above. I think we should remove that and just use `prod_single_zero` defined in `common_utils.py`

Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:1357]: Only used by a handful of functions like prod/cumprod. I decided it is better to reuse samples.

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:1357]: I don't understand. I was suggesting to remove https://github.com/pytorch/pytorch/pull/48125/files#diff-18a0c3e7cd07089b34760028c387d39de1aa393172f1d5f123692106b3a3d144R1344-R1348

**Review by anjali411 — APPROVED** [review]
LGTM and should be ready to be merged once we have resolved the discussion in our thread. thank you @nikitaved :D

File: torch/testing/_internal/common_methods_invocations.py
Comment (nikitaved) [reply torch/testing/_internal/common_methods_invocations.py:1357]: I am not sure it is very useful to use the version from utils and have it in utils because it is not being used:
```
(pytorch-cuda-dev) nik@qgpu3:~/git/pytorch$ git grep prod_single_zero
torch/testing/_internal/common_methods_invocations.py:    def prod_single_zero():
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput(prod_single_zero())
torch/testing/_internal/common_utils.py:def prod_single_zero(dim_size, device=torch.device('cpu'), dtype=torch.float):
(pytorch-cuda-dev) nik@qgpu3:~/git/pytorch$ git grep prod_zeros
torch/testing/_internal/common_methods_invocations.py:    def prod_zeros(dim_select):
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([0, 1]), 1))
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([0, 2]), 1))
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([1, 2]), 1))
torch/testing/_internal/common_methods_invocations.py:        yield SampleInput((prod_zeros([1, 2]), 1), kwargs={'dtype': dtype})
torch/testing/_internal/common_methods_invocations.py:def prod_zeros(dim_size, dim_select):

```

Comment (anjali411) [reply torch/testing/_internal/common_methods_invocations.py:1357]: oh I see what you are saying ... `prod_zeros` was only being used for `prod` and `cumprod` method_tests. ok so I think here's what we should do. let's keep the function defined above and let's remove the helper function in `common_utils.py`

**Comment by nikitaved** [comment]
Thank you, @anjali411 , the PR is ready.

**Comment by OrangeSodahub** [comment]
Has this issue been solved? I occurred this now.

**Comment by nikitaved** [comment]
@OrangeSodahub , what PyTorch version do you use? And what is your repro?

**Comment by OrangeSodahub** [comment]
@nikitaved torch1.12. repo is private.

**Comment by nikitaved** [comment]
@OrangeSodahub , I mean some minimal reproduction example. However, I cannot reproduce your modified example:

```python
In [1]: import torch

In [2]: torch.diag_embed(torch.rand(5, dtype=torch.cdouble)).det()
Out[2]: tensor(-0.0788+0.0172j, dtype=torch.complex128)

```

**Comment by OrangeSodahub** [comment]
@ngimel  Sorry. my version is `torch1.9+cu111`  and the log:
image

So weird, I cannot reproduce in another place.

**Comment by ngimel** [comment]
Complex support has been improved since 1.9, please open an issue with a minimal reproducible example if you are still seeing this error on the latest (1.13) release.

**Pull request closed without merging**
